archive/zchunk.ksy

```diff
@@ -0,0 +1,188 @@
+meta:
+  id: zchunk
+  title: Zchunk
+  file-extension: zck
+  license: CC0-1.0
+  endian: le
+doc-ref: https://github.com/zchunk/zchunk/blob/main/zchunk_format.txt
+seq:
+  - id: lead
+    type: lead
+  - id: header
+    type: header
+    size: lead.len_header.value
+  - id: chunks
+    size: chunk_metadata[_index].len_chunk.value
+    repeat: expr
+    repeat-expr: num_chunks
+instances:
+  num_chunks:
+    value: header.index.num_chunks.value - 1
+    doc: the number of chunks includes the header, so -1
+  chunk_metadata:
+    value: header.index.chunks_metadata
+types:
+  lead:
+    seq:
+      - id: magic
+        contents: [0, 'ZCK1']
+      - id: checksum
+        type: compressed_integer
+      - id: len_header
+        type: compressed_integer
+      - id: lead_checksum
+        size: len_checksum
+    instances:
+      checksum_type:
+        value: checksum.value
+        enum: checksum_types
+      len_checksum:
+        value: |
+            checksum_type == checksum_types::sha1 ? 20 :
+            checksum_type == checksum_types::sha256 ? 32 :
+            checksum_type == checksum_types::sha512 ? 64 :
+            checksum_type == checksum_types::sha512_128 ? 16 :
+            0
+  header:
+    seq:
+      - id: preface
+        type: preface
+      - id: len_index
+        type: compressed_integer
+      - id: index
+        type: index
+        size: len_index.value
+      - id: signatures
+        type: signatures
+  preface:
+    seq:
+      - id: checksum
+        size: _root.lead.len_checksum
+      - id: flags
+        type: compressed_integer
+      - id: compression_type
+        type: compressed_integer
+      - id: num_optional_element
+        type: compressed_integer
+        if: has_optional_elements
+      - id: optional_elements
+        type: optional_element
+        repeat: expr
+        repeat-expr: num_optional_element.value
+        if: has_optional_elements
+    instances:
+      has_data_streams:
+        value: flags.value & 0b1 == 0b1
+      has_optional_elements:
+        value: flags.value & 0b10 == 0b10
+  optional_element:
+    seq:
+      - id: element_id
+        type: compressed_integer
+      - id: len_data
+        type: compressed_integer
+      - id: data
+        size: len_data.value
+  index:
+    seq:
+      - id: checksum
```

Comment:
We can probably implement types calculating hashes purely in KS. Should we do it?